### ISSUE TITLE
TypeError on WooCommerce Single-Product View 

### ISSUE THREAD

**Issue by EhrhardtDa** [timber/timber#2657]
# Expected behavior
The single-produc.twig view works

## Actual behavior
The single-product.twig template results in an error:

```
2022-10-18T20:31:40+00:00 CRITICAL Uncaught TypeError: array_slice(): Argument ($length) must be of type ?int, array given in /Local Sites/dev/app/public/wp-content/plugins/woocommerce/includes/wc-template-functions.php:2087
Stack trace:
/Local Sites/dev/app/public/wp-content/plugins/woocommerce/includes/wc-template-functions.php(2087): array_slice(Array, 0, Array)
/Local Sites/dev/app/public/wp-includes/class-wp-hook.php(307): woocommerce_upsell_display(Array)
/Local Sites/dev/app/public/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
/Local Sites/dev/app/public/wp-includes/plugin.php(476): WP_Hook->do_action(Array)
[internal function]: do_action('woocommerce_aft...', Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/timber/timber/lib/Twig.php(39): call_user_func_array('do_action', Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Environment.php(418) : eval()'d code(82): Timber\Twig->Timber\{closure}(Array, 'woocommerce_aft...')
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(182): __TwigTemplate_3e4eacb3c48c6abfe4d23130ee6f9b06342cc4ac75be9f91c4f1a58584f0e93a->block_content(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Environment.php(418) : eval()'d code(75): Twig\Template->displayBlock('content', Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(405): __TwigTemplate_5e2fd50bace08ad92b25c44d0c07de13bb6bcaf5a154cda82fdd4d5de1ae8b28->doDisplay(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Environment.php(418) : eval()'d code(42): Twig\Template->display(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(405): __TwigTemplate_3e4eacb3c48c6abfe4d23130ee6f9b06342cc4ac75be9f91c4f1a58584f0e93a->doDisplay(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/Template.php(390): Twig\Template->display(Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render(Array, Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/timber/timber/lib/Loader.php(79): Twig\TemplateWrapper->render(Array)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/timber/timber/lib/Timber.php(334): Timber\Loader->render('views/woo/singl...', Array, false, 'default')
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/timber/timber/lib/Timber.php(383): Timber\Timber::compile('views/woo/singl...', Array, false, 'default', true)
/Local Sites/dev/app/public/wp-content/themes/storefront/vendor/timber/timber/lib/Timber.php(410): Timber\Timber::fetch('views/woo/singl...', Array, false, 'default')
/Local Sites/dev/app/public/wp-content/themes/storefront/woocommerce.php(25): Timber\Timber::render('views/woo/singl...', Array)
/Local Sites/dev/app/public/wp-includes/template-loader.php(106): include('/Users/davidehr...')
/Local Sites/dev/app/public/wp-blog-header.php(19): require_once('/Users/davidehr...')
/Local Sites/dev/app/public/index.php(17): require('/Users/davidehr...')
{main}
  thrown in /Local Sites/dev/app/public/wp-content/plugins/woocommerce/includes/wc-template-functions.php on line 2087
```

## Steps to reproduce behavior
Blank Setup with Wordpress & WooCommerce using the start theme and adding woocommerce support as described here:
https://timber.github.io/docs/guides/woocommerce/

## What version of WordPress, PHP and Timber are you using?
Wordpress 6.0.3, MySQL 8.0.16, PHP 8.1.9, Timber 1.21

## How did you install Timber? (for example, from GitHub, Composer/Packagist, WP.org?)
composer

**Comment by Leobaillard** [comment]
Same issue here with similar versions but with php 8.0.

**Comment by jaytuckwell** [comment]
We've just encountered the error on:

- Twig 1.21
- PHP 8.0
- WordPress 6.1

Any word on a fix?

**Comment by gchtr** [comment]
Could somebody help by boiling this down to the exact line of code in the Twig template that causes this error?

**Comment by jaytuckwell** [comment]
Thanks for the quick reply @gchtr

So I've just done some quick tests.

I can view the page after removing:

{% do action('woocommerce_after_single_product_summary') %}

Which I guess would correlate with the reference to `woocommerce_upsell_display` in the above logs, as this action calls; woocommerce_output_product_data_tabs, woocommerce_upsell_display & woocommerce_output_related_products.

**Comment by gchtr** [comment]
Could you replace

```twig
{% do action('woocommerce_after_single_product_summary') %}
```

with the following line and see if that fixes the issue? 

```twig
{% fn('do_action', 'woocommerce_after_single_product_summary') %}
```

**Comment by jaytuckwell** [comment]
Sure, I just tried with a small adjustment ( `{{ }}` vs `{% %}` )

It works if I use:

```
{{ fn('do_action', 'woocommerce_after_single_product_summary') }}
```

Thanks!

**Comment by gchtr** [comment]
@jaytuckwell `{{ }}` is only needed if the action wouldn’t already echo its result, but maybe only return a string.

This is an issue that we solved in #1773, so it will be fixed in Timber version 2. The problem is that Timber passes an additional parameter to the function behind the scenes when using `{% do action() %}`. So for Timber 1.x, this won’t be fixed.

We can work around this by using `{% fn('do_action', 'action') %}` instead of `{% do action('action') %}` in the meantime.

If somebody wants to pick this up and fix it in the documentation for v1, they’re welcome to add a pull request.
